File: app/src/main/java/vn/ditagis/com/tanhoa/qlsc/services/GetThietBi.java
package vn.ditagis.com.tanhoa.qlsc.services;

import android.content.Context;
import android.os.AsyncTask;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.data.Feature;
import com.esri.arcgisruntime.data.FeatureQueryResult;
import com.esri.arcgisruntime.data.QueryParameters;
import com.esri.arcgisruntime.data.ServiceFeatureTable;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ExecutionException;

import vn.ditagis.com.tanhoa.qlsc.entities.Constant;
import vn.ditagis.com.tanhoa.qlsc.entities.DLayerInfo;
import vn.ditagis.com.tanhoa.qlsc.entities.ThietBi;
import vn.ditagis.com.tanhoa.qlsc.entities.VatTu;
import vn.ditagis.com.tanhoa.qlsc.entities.entitiesDB.ListObjectDB;


public class GetThietBi extends AsyncTask<Void, Boolean, Void> {
    private Context mContext;
    private AsyncResponse mDelegate;

    public interface AsyncResponse {
        void processFinish();
    }


    public GetThietBi(Context context, AsyncResponse delegate) {
        this.mContext = context;
        this.mDelegate = delegate;
    }


    private void getVatTuFromService() {

        String layerInfoThietBi = Constant.ID_SU_CO_THIET_BI_TABLE;

        for (DLayerInfo dLayerInfo : ListObjectDB.getInstance().getLstFeatureLayerDTG()) {
            if (dLayerInfo.getId().equals(layerInfoThietBi)) {
                final QueryParameters queryParameters = new QueryParameters();
                queryParameters.setWhereClause("1=1");
                String url = dLayerInfo.getUrl();
                if (!url.startsWith("http"))
                    url = "http:" + dLayerInfo.getUrl();
                final ServiceFeatureTable serviceFeatureTable = new ServiceFeatureTable(url);
//
                final ListenableFuture<FeatureQueryResult> feature = serviceFeatureTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL);
                feature.addDoneListener(() -> {
                    final List<ThietBi> thietBiList = new ArrayList<>();
                    try {
                        FeatureQueryResult result = feature.get();
                        Iterator<Feature> iterator = result.iterator();
                        Feature item;

                        while (iterator.hasNext()) {
                            item = iterator.next();
                            String maThietBi = (String) item.getAttributes().get(Constant.FIELD_THIETBI.MA_THIET_BI);
                            String tenThietBi = (String) item.getAttributes().get(Constant.FIELD_THIETBI.TEN_THIET_BI);
                            ThietBi thietBi = new ThietBi(maThietBi, tenThietBi);
                            thietBiList.add(thietBi);
                        }
                        ListObjectDB.getInstance().setThietBis(thietBiList);
                        publishProgress(true);

                    } catch (InterruptedException | ExecutionException e) {
                        e.printStackTrace();
                        publishProgress();
                    }


                });

                break;
            }
        }

    }


    @Override
    protected Void doInBackground(Void... voids) {
        getVatTuFromService();
        return null;
    }

    @Override
    protected void onProgressUpdate(Boolean... values) {
        super.onProgressUpdate(values);
        if (values != null && values.length > 0 && values[0])
            mDelegate.processFinish();
    }
}

File: app/src/main/java/vn/ditagis/com/tanhoa/qlsc/services/GetThietBi.kt
package vn.ditagis.com.tanhoa.qlsc.services

import android.os.AsyncTask

import com.esri.arcgisruntime.data.Feature
import com.esri.arcgisruntime.data.QueryParameters
import com.esri.arcgisruntime.data.ServiceFeatureTable
import vn.ditagis.com.tanhoa.qlsc.entities.*

import java.util.ArrayList
import java.util.concurrent.ExecutionException

import vn.ditagis.com.tanhoa.qlsc.entities.entitiesDB.ListObjectDB


class GetThietBi(private val mApplication: DApplication, private val mDelegate: AsyncResponse) : AsyncTask<Void, Boolean, Void>() {

    interface AsyncResponse {
        fun processFinish()
    }


    private fun getVatTuFromService() {

        val layerInfoThietBi = Constant.IDLayer.ID_SU_CO_THIET_BI_TABLE

        for (dLayerInfo in mApplication.lstFeatureLayerDTG!!) {
            if (dLayerInfo.id == layerInfoThietBi) {
                val queryParameters = QueryParameters()
                queryParameters.whereClause = "1=1"
                var url = dLayerInfo.url
                if (!url.startsWith("http"))
                    url = "http:" + dLayerInfo.url
                val serviceFeatureTable = ServiceFeatureTable(url)
                //
                val feature = serviceFeatureTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL)
                feature.addDoneListener {
                    val thietBiList = ArrayList<ThietBi>()
                    try {
                        val result = feature.get()
                        val iterator = result.iterator()
                        var item: Feature

                        while (iterator.hasNext()) {
                            item = iterator.next()
                            val maThietBi = item.attributes[Constant.FieldThietBi.MA_THIET_BI] as String
                            val tenThietBi = item.attributes[Constant.FieldThietBi.TEN_THIET_BI] as String
                            val thietBi = ThietBi(maThietBi, tenThietBi)
                            thietBiList.add(thietBi)
                        }
                        ListObjectDB.instance.thietBis = thietBiList
                        publishProgress(true)

                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                        publishProgress()
                    } catch (e: ExecutionException) {
                        e.printStackTrace()
                        publishProgress()
                    }
                }
                break
            }
        }

    }


    override fun doInBackground(vararg voids: Void): Void? {
        getVatTuFromService()
        return null
    }

    override fun onProgressUpdate(vararg values: Boolean?) {
        super.onProgressUpdate(*values)
        if (values.isNotEmpty() && values[0]!!)
            mDelegate.processFinish()
    }
}
